### PR TITLE
[training] fix: validate MXFP8 param gather overlap

### PR DESCRIPTION
**Pull request by Glitchfix**
# What does this PR do ?

Adds validation so the non-FSDP MXFP8 parameter-gather path requires `ddp.overlap_param_gather=True` when `reuse_grad_buf_for_mxfp8_param_ag=True` is enabled.

# Changelog

- Require `ddp.overlap_param_gather=True` with non-FSDP MXFP8 gradient-buffer reuse.
- Extend config validation coverage for the invalid overlap-disabled case and the valid overlap-enabled case.

# GitHub Actions CI

Local validation:

- `uvx pre-commit run --files src/megatron/bridge/training/config.py tests/unit_tests/training/test_config.py`
- `PYTHONPATH=src:3rdparty/Megatron-LM python -m pytest tests/unit_tests/training/test_config.py::TestConfigContainerValidation::test_reuse_grad_buf_for_mxfp8_param_ag_required_without_fsdp`

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation? Not needed for this validation-only fix.
- [x] Does the PR affect components that are optional to install? No.

# Additional Information

- Related to #3801

### CODE REVIEW

**Comment by Glitchfix** [comment]
Requested labels per CONTRIBUTING.md: `bug`, `area:training`, `area:quant`, `needs-review`. I attempted to apply them directly, but this account does not have label permissions on the upstream repository.

**Comment by yaoyu-33** [comment]
Thanks for the fix. One question before we make this a hard validation: did you see a concrete failure when running with `fp8_param_gather=True`, `fp8_recipe='mxfp8'`, `reuse_grad_buf_for_mxfp8_param_ag=True`, and `ddp.overlap_param_gather=False`?

I checked the corresponding MCore handling and it looks like overlap is not required there: the overlap path does the param-buffer copy in the training loop, while the non-overlap path has separate handling in DDP/optimizer to copy gathered MXFP8 params and zero the shared buffer. MCore config validation only requires reuse to imply `fp8_param_gather`; it does not require `overlap_param_gather`.

If this combination is failing specifically in Bridge, that sounds like a separate Bridge-side bug from the existing `reuse_grad_buf_for_mxfp8_param_ag` requirement. Can you share the failure mode or repro that motivated requiring `ddp.overlap_param_gather=True` here?

File: src/megatron/bridge/training/config.py [hunk 1]
```diff
@@ -1238,6 +1238,10 @@ def validate(self) -> None:
                 "When fp8_param_gather=True and fp8_recipe='mxfp8', "
                 "reuse_grad_buf_for_mxfp8_param_ag must be set to True"
             )
+            assert self.ddp.overlap_param_gather, (
+                "When fp8_param_gather=True, fp8_recipe='mxfp8', and "
+                "reuse_grad_buf_for_mxfp8_param_ag=True, ddp.overlap_param_gather must be set to True"
+            )
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply src/megatron/bridge/training/config.py:1244]: @Glitchfix  can you check this comment?
